src/solvers/dg_structured/dg_3d.jl

```diff
@@ -0,0 +1,293 @@
+
+  return nothing
+end
+
+
+function calc_volume_integral!(du::AbstractArray{<:Any,5}, u,
+                               nonconservative_terms::Val{false}, mesh::StructuredMesh, equations,
+                               volume_integral::VolumeIntegralWeakForm,
+                               dg::DGSEM, cache)
+  @unpack derivative_dhat = dg.basis
+
+  @threaded for element in eachelement(dg, cache)
+    for k in eachnode(dg), j in eachnode(dg), i in eachnode(dg)
+      u_node = get_node_vars(u, equations, dg, i, j, k, element)
+
+      flux1 = transformed_flux(u_node, 1, mesh, equations)
+      for ii in eachnode(dg)
+        integral_contribution = derivative_dhat[ii, i] * flux1
+        add_to_node_vars!(du, integral_contribution, equations, dg, ii, j, k, element)
+      end
+
+      flux2 = transformed_flux(u_node, 2, mesh, equations)
+      for jj in eachnode(dg)
+        integral_contribution = derivative_dhat[jj, j] * flux2
+        add_to_node_vars!(du, integral_contribution, equations, dg, i, jj, k, element)
+      end
+
+      flux3 = transformed_flux(u_node, 3, mesh, equations)
+      for kk in eachnode(dg)
+        integral_contribution = derivative_dhat[kk, k] * flux3
+        add_to_node_vars!(du, integral_contribution, equations, dg, i, j, kk, element)
+      end
+    end
+  end
+
+  return nothing
+end
+
+
+function prolong2interfaces!(cache, u::AbstractArray{<:Any,5}, mesh::StructuredMesh, equations, dg::DG)
+  @threaded for element_ind in eachelement(dg, cache)
+    element = cache.elements[element_ind]
+
+    # Interfaces in x-direction
+    for k in eachnode(dg), j in eachnode(dg), v in eachvariable(equations)
+      element.interfaces[1].u_right[v, j, k] = u[v, 1, j, k, element_ind]
+      element.interfaces[2].u_left[v, j, k] = u[v, end, j, k, element_ind]
+    end
+
+    # Interfaces in y-direction
+    for k in eachnode(dg), i in eachnode(dg), v in eachvariable(equations)
+      element.interfaces[3].u_right[v, i, k] = u[v, i, 1, k, element_ind]
+      element.interfaces[4].u_left[v, i, k] = u[v, i, end, k, element_ind]
+    end
+
+    # Interfaces in z-direction
+    for j in eachnode(dg), i in eachnode(dg), v in eachvariable(equations)
+      element.interfaces[5].u_right[v, i, j] = u[v, i, j, 1, element_ind]
+      element.interfaces[6].u_left[v, i, j] = u[v, i, j, end, element_ind]
+    end
+  end
+
+  return nothing
+end
+
+
+function prolong2boundaries!(cache, u::AbstractArray{<:Any,5}, 
+    boundary_condition::BoundaryConditionPeriodic, mesh::StructuredMesh, equations, dg::DG)
+  @unpack size, linear_indices = mesh
+
+  # Boundaries in x-direction
+  for element_y in 1:size[2], element_z in 1:size[3]
+    # TODO Performance
+    cache.elements[1, element_y, element_z].interfaces[1].u_left .= u[:, end, :, :, linear_indices[end, element_y, element_z]]
+    cache.elements[end, element_y, element_z].interfaces[2].u_right .= u[:, 1, :, :, linear_indices[1, element_y, element_z]]
+  end
```

Comment:
We tried to make this here more efficient, but didn't really succeed.
This version allocates a lot. Adding `@views` everywhere reduced the allocations a lot, but made the code slower.
Replacing this with a loop over all variables and nodes removed all allocations, but was even slower.

How can we improve performance here?

Reply:
I am not sure I can reproduce your findings. To benchmark, I tried the following:
```julia
julia> trixi_include("examples/3d/elixir_advection_basic_structured.jl")

julia> u0_ode = similar(sol.u[begin])

julia> u0_ode .= 0

julia> u = Trixi.wrap_array(u0_ode, semi)

julia> using BenchmarkTools

julia> @benchmark Trixi.prolong2boundaries!($(semi.cache), $(u), $(boundary_condition_periodic), $(mesh), $(equations), $(solver))
```

Current implementation:
```julia
julia> @benchmark Trixi.prolong2boundaries!($(semi.cache), $(u), $(boundary_condition_periodic), $(mesh), $(equations), $(solver))
BenchmarkTools.Trial: 
  memory estimate:  78.00 KiB
  allocs estimate:  384
  --------------
  minimum time:     29.611 μs (0.00% GC)
  median time:      31.951 μs (0.00% GC)
  mean time:        44.121 μs (12.66% GC)
  maximum time:     5.158 ms (99.06% GC)
  --------------
  samples:          10000
  evals/sample:     1

```

With `@views` before every line in `prolong2boundaries!` (src/solvers/dg_structured/dg_3d.jl:111-135):
```julia
julia> @benchmark Trixi.prolong2boundaries!($(semi.cache), $(u), $(boundary_condition_periodic), $(mesh), $(equations), $(solver))
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     16.851 μs (0.00% GC)
  median time:      17.210 μs (0.00% GC)
  mean time:        17.270 μs (0.00% GC)
  maximum time:     86.482 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

```

Thus it seems to be faster by a factor of ~2 and allocation free.Can you please try to reproduce this yourself?

Reply:
Btw, I get similar results (but less of a speedup) for the Euler source terms test.

Reply:
Using Julia 1.6.0-rc3 (and 1.6.0-rc1 too), the version with `@views` was 3 times slower than the original version (running the exact same code as you).
In Julia 1.5.3, I got the same results as you.
I tried the new release candidates after I got different L2 and Linf errors in some Trixi examples, but I should probably just reinstall my Julia 1.5 version.